crucible-go/src/Lang/Crucible/Go/TransUtil.hs

```diff
@@ -0,0 +1,756 @@
+        _ -> Left (Some repr, Some e)
+
+asTypeMaybe :: TypeRepr b -> Gen.Expr Go s a -> Maybe (Gen.Expr Go s b)
+asTypeMaybe repr e = case asTypeEither repr e of
+  Left _x -> Nothing
+  Right e' -> Just e'
+
+asType :: TypeRepr b -> Gen.Expr Go s a -> Gen.Expr Go s b
+asType repr e =
+  case asTypeMaybe repr e of
+    Just e' -> e'
+    Nothing -> error $ "asType fail: expression " ++ show e ++ " of type "
+               ++ show (exprType e) ++ " incompatible with type " ++ show repr
+
+-- | CPS version of asType
+asType' :: TypeRepr b -> Gen.Expr Go s a -> (Gen.Expr Go s b -> c) -> c
+asType' repr e k = k $ asType repr e
+
+-- | asTypeEither lifted to assignments.
+asTypesEither :: CtxRepr ctx'
+              -> Assignment (Gen.Expr Go s) ctx
+              -> Either (Some TypeRepr, Some (Gen.Expr Go s))
+              (Assignment (Gen.Expr Go s) ctx')
+asTypesEither Empty Empty = Right Empty
+asTypesEither (reprs :> repr) (es :> e) =
+  pure (:>) <*> asTypesEither reprs es <*> asTypeEither repr e
+
+-- | asTypeMaybe lifted to assignments.
+asTypesMaybe :: CtxRepr ctx' ->
+                Assignment (Gen.Expr Go s) ctx ->
+                Maybe (Assignment (Gen.Expr Go s) ctx')
+asTypesMaybe Empty Empty = Just Empty
+asTypesMaybe (reprs :> repr) (es :> e) =
+  pure (:>) <*> asTypesMaybe reprs es <*> asTypeMaybe repr e
+
+asTypes :: CtxRepr ctx' ->
+           Assignment (Gen.Expr Go s) ctx ->
+           Assignment (Gen.Expr Go s) ctx'
+asTypes ctx assignment = case asTypesEither ctx assignment of
+  Right assignment' -> assignment'
+  Left (Some repr, Some e)->
+    error $ "asTypes: " ++ show e ++ " incompatible with " ++ show repr
+
+-- | CPS version of asTypes
+asTypes' :: CtxRepr ctx' ->
+            Assignment (Gen.Expr Go s) ctx ->
+            (Assignment (Gen.Expr Go s) ctx' -> a) -> a
+asTypes' ctxRepr assignment k = k $ asTypes ctxRepr assignment
+
+tryAsBV :: Gen.Expr Go s tp
+        -> (forall w. (1 <= w) => NatRepr w -> Gen.Expr Go s (BVType w) -> b)
+        -> b
+        -> b
+tryAsBV e k b = case exprType e of
+  BVRepr w -> k w e
+  _repr -> b
+
+tryAsStringRepr :: TypeRepr tp
+                -> (forall si. StringInfoRepr si ->
+                     TypeRepr (StringType si) -> b)
+               -> b
+               -> b
+tryAsStringRepr repr k b = case repr of
+  StringRepr si -> k si repr
+  _repr -> b
+
+tryAsString :: Gen.Expr Go s tp
+            -> (forall si. StringInfoRepr si ->
+                 Gen.Expr Go s (StringType si) -> b)
+            -> b
+            -> b
+tryAsString e k b = case exprType e of
+  StringRepr si -> k si e
+  _repr -> b
+
+tryAsPointer :: Gen.Expr Go s tp
+             -> (forall a. TypeRepr a -> Gen.Expr Go s (PointerType a) -> b)
+             -> b
+             -> b
+tryAsPointer e k b = case exprType e of    
+  MaybeRepr (VariantRepr
+             (Ctx.Empty :> ReferenceRepr repr :> StructRepr
+               (Ctx.Empty :> ReferenceRepr (VectorRepr repr') :> NatRepr))) ->
```

Comment:
Later tech debt note: I think there is a `PointerRepr` defined in the Encodings module that might make this cleaner.  Or perhaps that pattern synonym could be improved to make this easier (e.g., like the LLVMPointerRepr)

Reply:
Comment added.